// @nguniversal/common/esm5/common.es5.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { HTTP_INTERCEPTORS, HttpHeaders, HttpResponse } from '@angular/common/http';
import { ApplicationRef, Injectable, NgModule } from '@angular/core';
import { BrowserTransferStateModule, TransferState, makeStateKey } from '@angular/platform-browser';
import { of } from 'rxjs/observable/of';
import { filter } from 'rxjs/operators/filter';
import { tap } from 'rxjs/operators/tap';
import { take } from 'rxjs/operators/take';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} headers
 * @return {?}
 */
function getHeadersMap(headers) {
    var /** @type {?} */ headersMap = {};
    for (var _i = 0, _a = headers.keys(); _i < _a.length; _i++) {
        var key = _a[_i];
        headersMap[key] = /** @type {?} */ ((headers.getAll(key)));
    }
    return headersMap;
}
var TransferHttpCacheInterceptor = /** @class */ (function () {
    function TransferHttpCacheInterceptor(appRef, transferState) {
        var _this = this;
        this.transferState = transferState;
        this.isCacheActive = true;
        // Stop using the cache if the application has stabilized, indicating initial rendering is
        // complete.
        appRef.isStable
            .pipe(filter(function (isStable) { return isStable; }), take(1)).toPromise()
            .then(function () { _this.isCacheActive = false; });
    }
    /**
     * @param {?} url
     * @return {?}
     */
    TransferHttpCacheInterceptor.prototype.invalidateCacheEntry = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.transferState.remove(makeStateKey('G.' + url));
        this.transferState.remove(makeStateKey('H.' + url));
    };
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    TransferHttpCacheInterceptor.prototype.intercept = /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        // Stop using the cache if there is a mutating call.
        if (req.method !== 'GET' && req.method !== 'HEAD') {
            this.isCacheActive = false;
            this.invalidateCacheEntry(req.url);
        }
        if (!this.isCacheActive) {
            // Cache is no longer active. Pass the request through.
            return next.handle(req);
        }
        var /** @type {?} */ key = (req.method === 'GET' ? 'G.' : 'H.') + req.url;
        var /** @type {?} */ storeKey = makeStateKey(key);
        if (this.transferState.hasKey(storeKey)) {
            // Request found in cache. Respond using it.
            var /** @type {?} */ response = this.transferState.get(storeKey, /** @type {?} */ ({}));
            return of(new HttpResponse({
                body: response.body,
                headers: new HttpHeaders(response.headers),
                status: response.status,
                statusText: response.statusText,
                url: response.url,
            }));
        }
        else {
            // Request not found in cache. Make the request and cache it.
            var /** @type {?} */ httpEvent = next.handle(req);
            return httpEvent
                .pipe(tap(function (event) {
                if (event instanceof HttpResponse) {
                    _this.transferState.set(storeKey, {
                        body: event.body,
                        headers: getHeadersMap(event.headers),
                        status: event.status,
                        statusText: event.statusText,
                        url: /** @type {?} */ ((event.url)),
                    });
                }
            }));
        }
    };
    TransferHttpCacheInterceptor.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    TransferHttpCacheInterceptor.ctorParameters = function () { return [
        { type: ApplicationRef, },
        { type: TransferState, },
    ]; };
    return TransferHttpCacheInterceptor;
}());
/**
 * An NgModule used in conjunction with `ServerTransferHttpCacheModule` to transfer cached HTTP
 * calls from the server to the client application.
 */
var TransferHttpCacheModule = /** @class */ (function () {
    function TransferHttpCacheModule() {
    }
    TransferHttpCacheModule.decorators = [
        { type: NgModule, args: [{
                    imports: [BrowserTransferStateModule],
                    providers: [
                        TransferHttpCacheInterceptor,
                        { provide: HTTP_INTERCEPTORS, useExisting: TransferHttpCacheInterceptor, multi: true },
                    ],
                },] },
    ];
    /** @nocollapse */
    TransferHttpCacheModule.ctorParameters = function () { return []; };
    return TransferHttpCacheModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// export { StateTransferInitializerModule } from './src/state-transfer-initilizer/module';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { TransferHttpCacheModule, TransferHttpCacheInterceptor as ɵa };
